File: app/src/main/java/top/pythong/noteblog/data/RestResponse.java
package top.pythong.noteblog.data;


/**
* @author ChangJiahong
* @date 2019/7/16
*/
public class RestResponse<T> {

    // 响应业务状态
    private Integer status;

    // 响应消息
    private String msg;

    // 响应中的数据
    private T data;

    private Boolean ok;

    /**
     * 服务器响应时间
     */
    private long timestamp;

    /**
     * 两个参数
     * status、msg
     * @param status
     * @param msg
     */
    private RestResponse(int status, String msg){
        this(status, msg, null);
    }

    /**
     * 三个参数
     * @param status
     * @param msg
     * @param data
     */
    private RestResponse(int status, String msg, T data) {
        this.status = status;
        this.msg = msg;
        this.data = data;
        this.timestamp = System.currentTimeMillis() / 1000;
    }


    public Boolean getOk() {
        return ok;
    }

    public void setOk(Boolean ok) {
        this.ok = ok;
    }

    public Integer getStatus() {
        return status;
    }

    public void setStatus(Integer status) {
        this.status = status;
    }

    public String getMsg() {
        return msg;
    }

    public void setMsg(String msg) {
        this.msg = msg;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }



}

File: app/src/main/java/top/pythong/noteblog/data/RestResponse.kt
package top.pythong.noteblog.data


/**
 * 返回viewModel层的数据类
 *
 * @author ChangJiahong
 * @date 2019/7/16
 */
data class RestResponse<T>(
    val ok: Boolean,
    val status: Int,
    val msg: String?,
    val data: T?,
    val timestamp: Long
) {

    constructor(ok: Boolean, status: Int, msg: String, data: T?) : this(
        ok, status, msg,
        data, 0L
    )

    private constructor(restResponse: RestResponse<*>) : this(
        restResponse.ok, restResponse.status,
        restResponse.msg?:"", restResponse.data as T
    )

    fun <T> convertTo(tClass: Class<T>): RestResponse<T> {
        return RestResponse(this)
    }

    fun isOk(): Boolean {
        return ok
    }

    companion object {
        fun <T> fail(status: Int, msg: String): RestResponse<T> {
            return RestResponse(false, status, msg, null)
        }
    }
}
